hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/StoreFileWriter.java

```diff
@@ -256,156 +251,560 @@ public void appendMobMetadata(SetMultimap<TableName, String> mobRefSet) throws I
+      ) {
+        return true;
+      }
     }
-    return haveBloom;
+    return false;
   }
 
-  private boolean closeGeneralBloomFilter() throws IOException {
-    boolean hasGeneralBloom = closeBloomFilter(generalBloomFilterWriter);
+  private boolean isDeletedByDeleteColumn(Cell cell) {
+    return deleteColumn != null && (deleteColumn.getTimestamp() > cell.getTimestamp()
+      || (deleteColumn.getTimestamp() == cell.getTimestamp()
+        && (!newVersionBehavior || cell.getSequenceId() < deleteColumn.getSequenceId())));
+  }
 
-    // add the general Bloom filter writer and append file info
-    if (hasGeneralBloom) {
-      writer.addGeneralBloomFilter(generalBloomFilterWriter);
-      writer.appendFileInfo(BLOOM_FILTER_TYPE_KEY, Bytes.toBytes(bloomType.toString()));
-      if (bloomParam != null) {
-        writer.appendFileInfo(BLOOM_FILTER_PARAM_KEY, bloomParam);
+  private boolean isDeletedByDeleteColumnVersion(Cell cell) {
+    for (Cell deleteColumnVersion : deleteColumnVersionList) {
+      if (
+        deleteColumnVersion.getTimestamp() == cell.getTimestamp()
+          && (!newVersionBehavior || cell.getSequenceId() < deleteColumnVersion.getSequenceId())
+      ) {
+        return true;
       }
-      bloomContext.addLastBloomKey(writer);
     }
-    return hasGeneralBloom;
+    return false;
   }
 
-  private boolean closeDeleteFamilyBloomFilter() throws IOException {
-    boolean hasDeleteFamilyBloom = closeBloomFilter(deleteFamilyBloomFilterWriter);
+  private boolean isDeleted(Cell cell) {
+    return isDeletedByDeleteFamily(cell) || isDeletedByDeleteColumn(cell)
+      || isDeletedByDeleteFamilyVersion(cell) || isDeletedByDeleteColumnVersion(cell);
+  }
 
-    // add the delete family Bloom filter writer
-    if (hasDeleteFamilyBloom) {
-      writer.addDeleteFamilyBloomFilter(deleteFamilyBloomFilterWriter);
+  private void appendCell(Cell cell) throws IOException {
+    if ((lastCell == null || !CellUtil.matchingColumn(lastCell, cell))) {
+      initColumnState();
     }
+    if (cell.getType() == Cell.Type.DeleteFamily) {
+      if (deleteFamily == null) {
+        deleteFamily = cell;
+        liveFileWriter.append(cell);
+      } else {
+        getHistoricalFileWriter().append(cell);
+      }
+    } else if (cell.getType() == Cell.Type.DeleteFamilyVersion) {
+      if (!isDeletedByDeleteFamily(cell)) {
+        deleteFamilyVersionList.add(cell);
+        if (deleteFamily != null && deleteFamily.getTimestamp() == cell.getTimestamp()) {
+          // This means both the delete-family and delete-family-version markers have the same
+          // timestamp but the sequence id of delete-family-version marker is higher than that of
+          // the delete-family marker. In this case, there is no need to add the
+          // delete-family-version marker to the live version file. This case happens only with
+          // the new version behavior.
+          liveFileWriter.append(cell);
```

Comment:
Yes, the cell should have been added to the historical file. Please feel free to file a jira and fix it. It not a data integrity issue but it should be fixed. If you prefer me to do, please let me know. Good catch!

Reply:
Thank you. I'm glad to fix this issue. However, I would like to ask why there is no need to add the delete-family-version marker to the live version file here? Won't there be any correctness issues? The deletion order in the new version behavior itself takes the sequence id into account. Here, the sequence id of the delete-family-version marker is larger than that of the delete-family marker. Won't directly putting the delete-family-version marker into the historical file lead to the loss of deletion information? Thank you.

Reply:
If we have delete-family and delete-family-version markers with the same timestamp then the set of cells that are deleted by the delete-family-version marker is also covered (that is delete) by the delete-family marker. The reason we need delete markers in the live version file is to make sure that we do not return the cells previously stored in live files but deleted later would not be returned by the regular (not raw) scans that scan the latest versions of the cells. There might more than one delete marker that deletes a given cell. However, we need only one of them to mask this deleted cell during scans.

Reply:
Thank you. Yes, under the default version behavior, I'm sure that the delete-family marker can overwrite the delete-family-version marker. However, under the new version behavior, I'm not sure whether such a situation will occur: there is a delete-family marker with a timestamp (ts) of 10 and a sequence ID (seqId) of 100, a Put with a ts of 10 and a seqId of 101, and a delete-family-version marker with a ts of 10 and a seqId of 102. At this time, the Put is invisible. If during compaction, the delete-family-version marker is placed into the historical file, but the delete-family marker and the Put are in the live version file, then there will be a problem. Under the new version behavior, the Put will become visible. I would like to ask how this situation is avoided. During compaction, if the delete-family marker and the delete-family-version marker appear simultaneously, will the Put also definitely appear (so that the Put will definitely be placed into the historical file)? What kind of mechanism guarantees this? Thx!

Reply:
Thank you for being diligent! I think there is no bug here. Now it is clear to me there is no inconsistency between the comment and the code as opposed to what I thought initially. The reason we write this delete version marker to cover the case you mentioned above. I agree my comment could have been phrased better as follow.

```
          // This means both the delete-family and delete-family-version markers have the same
          // timestamp but the sequence id of delete-family-version marker is higher than that of
          // the delete-family marker. With the new version behavior, the delete-family-version 
          // maker is not deleted (masked) by the delete-family marker with the same timestamp 
          // as in the case here. That is why we need to write this delete marker to the live version
          // file.
```

Reply:
Thank you! But I think the handling of the Delete type is similar to this, but the Delete is placed in the historical file. The code is here: https://github.com/apache/hbase/blob/2531f9382018cf6548deb97fafedab4b102ab277/hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/StoreFileWriter.java#L432

Reply:
You are right! Let us fix that and rephrase the comments to make them clear. Also please check other cases too. I will review your PR.

Reply:
Okay, I'm glad to fix this problem. However, I want to confirm that the correct behavior should be adding the Delete operation here to the live version file, right?